// @adonisjs/ace/build/src/BaseCommand/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fancy_logs_1 = require("@poppinss/fancy-logs");
const colors_1 = require("@poppinss/colors");
const prompts_1 = require("@poppinss/prompts");
const Generator_1 = require("../Generator");
class BaseCommand {
    constructor(rawMode = false) {
        this.rawMode = rawMode;
        this.prompt = this.rawMode ? new prompts_1.FakePrompt() : new prompts_1.Prompt();
        this.logger = new fancy_logs_1.Logger({ fake: this.rawMode });
        this.generator = new Generator_1.Generator();
    }
    static boot() {
        if (this.booted) {
            return;
        }
        this.booted = true;
        Object.defineProperty(this, 'args', { value: [] });
        Object.defineProperty(this, 'flags', { value: [] });
        if (!this.hasOwnProperty('settings')) {
            Object.defineProperty(this, 'settings', { value: {} });
        }
        if (!this.hasOwnProperty('commandName')) {
            Object.defineProperty(this, 'commandName', { value: '' });
        }
        if (!this.hasOwnProperty('description')) {
            Object.defineProperty(this, 'description', { value: '' });
        }
    }
    get colors() {
        return this.rawMode ? new colors_1.FakeColors() : new colors_1.Colors();
    }
}
exports.BaseCommand = BaseCommand;
