// @adonisjs/ace/build/src/Decorators/flags.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const param_case_1 = __importDefault(require("param-case"));
function addFlag(type, options) {
    return function flag(target, propertyName) {
        const flag = Object.assign({
            name: param_case_1.default(propertyName),
            propertyName,
            type,
        }, options);
        target.constructor.boot();
        target.constructor.flags.push(flag);
    };
}
exports.flags = {
    string(options) {
        return addFlag('string', options || {});
    },
    number(options) {
        return addFlag('number', options || {});
    },
    boolean(options) {
        return addFlag('boolean', options || {});
    },
    array(options) {
        return addFlag('array', options || {});
    },
    numArray(options) {
        return addFlag('numArray', options || {});
    },
};
